### ISSUE TITLE
Cannot setup lsp

### ISSUE THREAD

**Issue by sudipghimire533** [NvChad/NvChad#1438]
**Describe the bug**
On fresh setup of LSP I cannot setup any lsp

**To Reproduce**
1. Freshly install NvChad
2. have following directory structure
```
/home/user/.config/nvim/
├── examples
│   ├── chadrc.lua
│   └── init.lua
├── init.lua
├── LICENSE
├── lua
│   ├── core
│   │   ├── default_config.lua
│   │   ├── init.lua
│   │   ├── lazy_load.lua
│   │   ├── mappings.lua
│   │   ├── options.lua
│   │   ├── packer.lua
│   │   └── utils.lua
│   ├── custom
│   │   ├── chadrc.lua
│   │   └── plugins
│   │       └── lspconfig.lua
│   └── plugins
│       ├── configs
│       │   ├── alpha.lua
│       │   ├── cmp.lua
│       │   ├── lspconfig.lua
│       │   ├── mason.lua
│       │   ├── nvimtree.lua
│       │   ├── nvterm.lua
│       │   ├── others.lua
│       │   ├── telescope.lua
│       │   ├── treesitter.lua
│       │   └── whichkey.lua
│       └── init.lua
└── plugin
    └── packer_compiled.lua

8 directories, 25 files
```
3. Following file content
**lua/custom/chadrc.lua**
```lua
-- Just an example, supposed to be placed in /lua/custom/

local M = {}

-- make sure you maintain the structure of `core/default_config.lua` here,
-- example of changing theme:

M.ui = {
}

M.plugins = {

  override = {
  },

  user = require "custom.plugins",
}

return M
```

**lua/custom/plugins/lspconfig.lua**
```lua
local on_attach = require("plugins.configs.lspconfig").on_attach
local capabilities = require("plugins.configs.lspconfig").capabilities

local lspconfig = require("lspconfig")
local servers = { "html", "cssls", "emmet_ls", "clangd", "jsonls", "tsserver" }

for _, lsp in ipairs(servers) do
	lspconfig[lsp].setup({
		on_attach = on_attach,
		capabilities = capabilities,
	})
end
```

**lua/custom/plugins/init.lua**
```lua
return {
  ["neovim/nvim-lspconfig"] = {
    config = function()
      require "plugins.configs.lspconfig"
      require "custom.plugins.lspconfig"
    end,
  },
}
```

4. Execute `:MasonInstall clangd`
5. Restart neovim
6. Execute `:LspInfo`

**Expected behavior**
Shows that clagd and other stuff have been configured but show only sumneko_lua is configured. I also tried to create a c++ project and edit a cpp file. No lsp feature whatsoever

**Screenshots**
![image](https://user-images.githubusercontent.com/39845901/183278685-9bcaa61a-0cca-49c8-8227-10bf2a9886ab.png)
![image](https://user-images.githubusercontent.com/39845901/183278693-416af03d-0b40-438b-a0e1-eded073d131e.png)

**Desktop (please complete the following information):**
 - Operating System: Debian Linux
 - Terminal: Mate-terminal
 - Version of Neovim: 0.7.2

**Comment by nullchilly** [comment]
In the meantime try this

```lua
return {
  ["neovim/nvim-lspconfig"] = {
    config = function()
      require "plugins.configs.lspconfig"
      require "custom.plugins.lspconfig"
      vim.cmd "silent e!"
    end,
  },
}
```

**Comment by tomasky** [comment]
> In the meantime try this In `lua/custom/plugins/init.lua`
> 
> ```lua
> return {
>   ["neovim/nvim-lspconfig"] = {
>     config = function()
>       require "plugins.configs.lspconfig"
>       require "custom.plugins.lspconfig"
>       vim.cmd "silent e!"
>     end,
>   },
> }
> ```

no  , ` vim.cmd "silent e!"`  is not necessary

@sudipghimire533  ,please provide your custom configs ,i can test it

**Comment by sudipghimire533** [comment]
Thank you so much for response:

Here it is: https://tmpfiles.org/3146/nvim.zip

**Comment by nullchilly** [comment]
@tomasky It is necessary because the `vim.defer_fn` messed up filetype detection. We used to have this in the past to cope with it

**Comment by tomasky** [comment]
this config is outdated

**Comment by nullchilly** [comment]
Bruh of course it is outdated

But my point is on my config if I disable lsp lazyloading it works correctly

And had to do this when defering nvim-lspconfig

**Comment by tomasky** [comment]
```lua
-- https://github.com/NvChad/NvChad/blob/main/lua/core/lazy_load.lua
M.lazy_load = function(tb)
  autocmd(tb.events, {
    group = vim.api.nvim_create_augroup(tb.augroup_name, {}),
    callback = function()
      if tb.condition() then
        vim.api.nvim_del_augroup_by_name(tb.augroup_name)

        -- dont defer for treesitter as it will show slow highlighting
        -- This deferring only happens only when we do "nvim filename"
        if tb.plugin ~= "nvim-treesitter" then
          vim.defer_fn(function()
            require("packer").loader(tb.plugin)
            if tb.plugin == "nvim-lspconfig" then
              vim.cmd "silent! do FileType"           --- here handler it already
            end
          end, 0)
        else
          require("packer").loader(tb.plugin)
        end
      end
    end,
  })
end

```

**Comment by nullchilly** [comment]
@tomasky You need to load the plugin first then `do FileType`

**Comment by tomasky** [comment]
@sudipghimire533  ,i use your custom configs ,and install `tsserver`  lsp using :Mason, it works normally.
So check your clangd ,see if it is installed correctly and  exec :PackerSync

**Comment by nullchilly** [comment]
@tomasky No, you has to install clangd to see it, not other language server...

**Comment by sudipghimire533** [comment]
> @sudipghimire533  ,i use your custom configs ,and install `tsserver`  lsp using :Mason, it works normally.
> So check your clangd ,see if it is installed correctly and  exec :PackerSync 

I tried with rust workspace as well. No progress. Let me try again anyway

**Comment by nullchilly** [comment]
Replacing
```lua
vim.cmd "silent! do FileType"
```
with
```lua
vim.cmd "silent! e %" end
```
fixed the issue.

**Comment by siduck** [comment]
i beg all the newbies, please read whole docs from the start to end sequentially

**Comment by nullchilly** [comment]
@siduck Did you care to read our conversation? lsp works on lua, tsserver

The issue is reporting it isn't work on clangd and rust

Your video only tested lua, tsserver

**Comment by tomasky** [comment]
i dont install clangd ,just install gopls and rust-analyzer , and they works also.

**Comment by sudipghimire533** [comment]
It turns out that clangd was not executable.

And about rust, It is working when I disable the tree-setter and not working when I re-enable it. I don't know what causing it but I will continue to work with treesetter disable for a while then

**Comment by sudipghimire533** [comment]
> @siduck Did you care to read our conversation? lsp works on lua, tsserver
> 
> The issue is reporting it isn't working on clangd and rust
> 
> Your video only tested lua, tsserver that is unrelated here
> 
> This is a `but it works on my end` card
> 
> There could be undefined behavior here with defer and filetype detection

Yep on my end as well I said lua was working fine. anyway for future reader if you want to ask question: you MUST watch youtube videos that's not mentioned elsewhere. Anyway this was first time I had issue with NvChad in last few months, hope I will find workaround.

Kudos to the creator

**Comment by tomasky** [comment]
> It turns out that clangd was not executable.

This indicates that clangd is not properly installed.

**Comment by sudipghimire533** [comment]
> > It turns out that clangd was not executable.
> 
> This indicates that clangd is not properly installed.

Yeah but what do you think might be the case with rust-analyzer

**Comment by tomasky** [comment]
> > > It turns out that clangd was not executable.
> > 
> > 
> > This indicates that clangd is not properly installed.
> 
> Yeah but what do you think might be the case with rust-analyzer

i dont know ,my tree-setter  is enabled on by default and my rust-analyzer works
maybe you can check this process after opening a rust file, it should output two processes
```bash
ps aux|grep rust-analyzer
```

**Comment by tomasky** [comment]
local servers = { "html", "cssls", "emmet_ls", "clangd", "jsonls", "tsserver" }

You dont config rust-analyzer?

**Comment by tomasky** [comment]
my lsp config
```lua
local servers = { "tsserver", "pyright", "rust_analyzer", "gopls", "nimls" }
```

**Comment by nullchilly** [comment]
@sudipghimire533 Maybe it's an issue with debian?

**Comment by sudipghimire533** [comment]
> @sudipghimire533 Maybe it's an issue with debian? What's the log when you do `./clangd`
```
└──╼ $clangd
clangd is a language server that provides IDE-like features to editors.

It should be used via an editor plugin rather than invoked directly. For more information, see:
	https://clangd.llvm.org/
	https://microsoft.github.io/language-server-protocol/

clangd accepts flags on the commandline, and in the CLANGD_FLAGS environment variable.

I[15:08:05.295] Debian clangd version 11.0.1-2
I[15:08:05.295] PID: 12063
I[15:08:05.295] Working directory: /home/sudip
I[15:08:05.295] argv[0]: clangd
I[15:08:05.295] Starting LSP over stdin/stdout

```

**Comment by tomasky** [comment]
no , Mason will install   its own clangd,check  dir:

~/.local/share/nvim/mason

**Comment by siduck** [comment]
> > > It turns out that clangd was not executable.
> > 
> > 
> > This indicates that clangd is not properly installed.
> 
> Yeah but what do you think might be the case with rust-analyzer

ig rust must be installed to the system

**Comment by siduck** [comment]
> @siduck Did you care to read our conversation? lsp works on lua, tsserver
> 
> The issue is reporting it isn't working on clangd and rust
> 
> Your video only tested lua, tsserver that is unrelated here
> 
> This is a `but it works on my end` card
> 
> There could be undefined behavior here with defer and filetype detection

this still isnt a nvchad issue, certain lsps require specific programs like clang etc on the system to be installed
